src/data/fintype/order.lean

```diff
@@ -0,0 +1,119 @@
+  .. (infer_instance : semilattice_sup α)}
+
+/-- An inhabited finite `lattice` is bounded  -/
+def fintype.bounded_lattice (α : Type*) [inhabited α] [fintype α] [lattice α] :
+  bounded_lattice α :=
+{ .. fintype.semilattice_inf_bot α,
+  .. fintype.semilattice_sup_top α}
+
+end to_bot_top
+
+section bot_top_eq
+
+local attribute [instance] fintype.semilattice_inf_bot
+
+lemma fintype.bot_eq {α : Type*} [inhabited α] [fintype α] [semilattice_inf α] :
+  ⊥ = finset.fold (⊓) (arbitrary α) id finset.univ :=
+rfl
+
+local attribute [instance] fintype.semilattice_sup_top
+
+lemma fintype.top_eq {α : Type*} [inhabited α] [fintype α] [semilattice_sup α] :
+  ⊤ = finset.fold (⊔) (arbitrary α) id finset.univ :=
+rfl
+
+end bot_top_eq
+
+section complete
+
+open_locale classical
+
+/-- A finite bounded lattice is complete -/
+noncomputable def fintype.complete_lattice (α : Type*) [fintype α] [h : bounded_lattice α] :
+  complete_lattice α :=
+{ Sup := λ s, s.to_finset.sup id,
+  Inf := λ s, s.to_finset.inf id,
+  le_Sup := λ _ _ ha, finset.le_sup (set.mem_to_finset.mpr ha),
+  Sup_le := λ s _ ha, finset.sup_le (λ b hb, ha _ (set.mem_to_finset.mp hb)),
+  Inf_le := λ _ _ ha, finset.inf_le (set.mem_to_finset.mpr ha),
+  le_Inf := λ s _ ha, finset.le_inf (λ b hb, ha _ (set.mem_to_finset.mp hb)),
+  ..h}
+
+/-- A nonempty finite lattice is complete. If the lattice is already a `bounded_lattice`, then use
+`fintype.complete_lattice` instead, as this gives definitional equality for `⊥` and `⊤` -/
+noncomputable def fintype.complete_lattice_of_lattice (α : Type*) [fintype α] [i : nonempty α]
+  [lattice α] : complete_lattice α :=
+@fintype.complete_lattice _ _ (@fintype.bounded_lattice α ⟨i.some⟩ _ _)
+
+/-- A nonempty finite linear order is complete.
+
+If the lattice is already a `bounded_lattice`, then build this object manually instead by combining
+`fintype.complete_lattice` and an appropriate `linear_order` instance, as this gives definitional
+equality for `⊥` and `⊤` -/
+noncomputable def fintype.complete_linear_order (α : Type*)
+  [fintype α] [nonempty α] [linear_order α] : complete_linear_order α :=
+{ .. fintype.complete_lattice_of_lattice α,
+  .. (infer_instance : linear_order α) }
+
+noncomputable instance fin.complete_linear_order {n : ℕ} : complete_linear_order (fin (n+1)) :=
+{ .. fintype.complete_lattice (fin (n+1)),
+  .. (infer_instance : linear_order (fin (n+1)))}
+
+end complete
+
+section decidable_Sup_Inf
+
+local attribute [instance] fintype.complete_lattice
+
+/-- The `Sup` induced by `fintype.complete_lattice` unfolds to `finset.sup`. -/
+lemma fintype.Sup_eq {α : Type*} [fintype α] [bounded_lattice α] (s : set α)
+  [decidable_pred (∈ s)] : Sup s = s.to_finset.sup id :=
+by convert rfl
+
+/-- The `Inf` induced by `fintype.complete_lattice` unfolds to `finset.inf`. -/
+lemma fintype.Inf_eq {α : Type*} [fintype α] [bounded_lattice α] (s : set α)
+  [decidable_pred (∈ s)] : Inf s = s.to_finset.inf id :=
+by convert rfl
```

Comment:
It would be nice if we could also prove:
```lean
/-- The `supr` induced by `fintype.complete_lattice` unfolds to `finset.sup`. -/
lemma fintype.supr_eq {α ι: Type*} [fintype ι] [fintype α] [bounded_lattice α] (f : ι → α) :
  supr f = finset.univ.sup f :=
begin
  classical,
  rw [supr, fintype.Sup_eq],
  have : (set.range f).to_finset = finset.univ.image f,
  { ext, simp?, },
  rw this,
  sorry,
end
```

Reply:
I need #7428 to prove this

Reply:
And also probably #7426 to avoid that `have`